### PR TITLE
feat: add object to subscribe notification callback

### LINKED ISSUE #19097: systemPreferences.subscribeNotification should pass notification object to the callback
### Preflight Checklist
* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for a feature request that matches the one I want to file, without success.

### Problem Description
Today, it's not possible to receive contextual data from notifications originating from sandboxed macOS apps.

This is because Electron's `systemPreferences.subscribeNotification` accepts a callback that receives the `event` and `userInfo` properties from the macOS notification but **not** the `object` property. This is problematic because sandboxed apps cannot use the `userInfo` property, and can only transmit contextual data via the `object` of a notification. [Apple's docs](https://developer.apple.com/documentation/foundation/nsdistributednotificationcenter/1418360-postnotificationname?language=objc) clearly note this restriction.

As such, Electron cannot read contextual data sent from notifications that originate from sandboxed apps. So, for example, a Finder extension can notify an Electron app that *some* file has been right-clicked, but **cannot** indicate *which file* has been right-clicked.

### Proposed Solution
I'm proposing a change similar to #5582, which enhanced `systemPreferences.subscribeNotification` to pass the `userInfo` from the macOS notification to the callback. Here, we need to further enhance `systemPreferences.subscribeNotification` to also pass the `object` from the macOS notification.

To preserve backwards-compatibility, the callback should accept the new `object` argument as its third parameter. As such, the new version of the documentation [here](https://electronjs.org/docs/api/system-preferences#systempreferencessubscribenotificationevent-callback-macos) would read:

```diff
systemPreferences.subscribeNotification(event, callback) [macOS]

* event String
* callback Function
    * event String
    * userInfo Object
+   * object String

Returns Number - The ID of this subscription

Subscribes to native notifications of macOS, callback will be called with `callback(event, userInfo)` when the corresponding event happens. The `userInfo` is an Object that contains the user information dictionary sent along with the notification.
+The `object` is a String that contains additional data sent along with the notification.
```

### Alternatives Considered
I think the only reasonable alternative is to update the documentation [here](https://electronjs.org/docs/api/system-preferences#systempreferencessubscribenotificationevent-callback-macos) to note that sandbox apps cannot send any contextual data to Electron apps.

### Additional Information
There is a precedent for this kind of change in Electron. See #5543 which was fixed under #5582.

For macOS developers running into this issue, a possible workaround is to have your sandboxed app communicate with a non-sandboxed app and then relay notifications on to your Electron app. The sandbox to non-sandbox communication must use `object` to hold contextual data, and the non-sandbox to Electron communication must use `userInfo` to hold contextual data. (However, AFAIK, using a non-sandboxed app prevents publishing your app in the Mac App Store).

### PR DESCRIPTION
**Pull request by louismrose**
#### Description of Change

Closes #19097.

This change exposes the `object` property on NSNotification to the callback passed to Electron's `systemPreferences.subscriptionNotification` methods.

I'm cc-ing @zcbenz on this PR as he worked on a very similar feature under #5582.

#### Checklist
- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Exposed the value of NSNotification.object to subscribers of notifications in systemPreferences.

### CODE REVIEW

**Comment by louismrose** [comment]
I'm not too sure why these builds are failing -- I don't think it's related to my changes, but I might be missing something obvious! Any tips from a contributor would be greatly appreciated? I'm happy to tackle any further changes needed here.

File: docs/api/system-preferences.md [hunk 1]
```diff
@@ -90,13 +90,14 @@ that contains the user information dictionary sent along with the notification.
...
   * `userInfo` Object
+  * `object` String
 
 Returns `Number` - The ID of this subscription
 
 Subscribes to native notifications of macOS, `callback` will be called with
 `callback(event, userInfo)` when the corresponding `event` happens. The
 `userInfo` is an Object that contains the user information dictionary sent
-along with the notification.
+along with the notification. The `object` is the sender of the notification.
```
Comment (zcbenz) [inline docs/api/system-preferences.md]: It seems that the `object` is only meaningful when it is a `NSString`?

I'm good with only supporting strings for now, but the documentation should explicitly document that "only `NSSString` is supported for now".

Comment (louismrose) [reply docs/api/system-preferences.md]: Ahh, yes, you're right. I hadn't spotted that other types might be used.

I've updated the docs to make this limitation clear.

Thanks for the helpful feedback!

**Comment by louismrose** [comment]
Thanks for the feedback @zcbenz. I've made the documentation change that you suggested.

**Comment by louismrose** [comment]
I'm pretty sure that the Mac build failure is a false positive, and that my changes have not introduced a regression. 

The error from the [CI build](https://circleci.com/gh/electron/electron/255764) is:

```
Error: No valid signing identity available to run autoUpdater specs
    at Context.<anonymous> (electron/spec-main/api-autoupdater-darwin-spec.ts:26:15)
    at processImmediate (internal/timers.js:445:21)
```

And this is for the `autoUpdater behavior "before each" hook for "should have a valid code signing identity"` part of the test suite.

It seems like perhaps this is a CI setup failure, rather than anything to do with the code I've changed (which is in system preferences, and not in auto-updating).

@zcbenz -- do you know if there's anything more I need to do to get a passing build, or for this to be merged? I looked at re-running the build on Circle CI, but I don't have permission to do so. The [Electron docs](https://electronjs.org/docs/development/pull-requests#continuous-integration-testing) suggest that I could ask a releaser to re-run the build, but sadly the link to the list of releasers ((https://github.com/orgs/electron/teams/releasers/members) is a 404 for me.

**Comment by zcbenz** [comment]
I have re-run the test, we can merge even with CI failing since it is obviously a flaky one.

(I want to wait for more reviews before merging this, as it is a new feature.)

**Comment by louismrose** [comment]
Great, thanks for your help @zcbenz!

File: shell/browser/api/atom_api_system_preferences_mac.mm [hunk 2]
```diff
@@ -212,11 +212,14 @@ AVMediaType ParseMediaType(const std::string& media_type) {
                     NSDictionaryToDictionaryValue(notification.userInfo);
                 if (user_info) {
                   copied_callback.Run(
-                      base::SysNSStringToUTF8(notification.name), *user_info);
+                      base::SysNSStringToUTF8(notification.name),
+                      *user_info,
+                      base::SysNSStringToUTF8(notification.object));
```
Comment (nornagon) [inline shell/browser/api/atom_api_system_preferences_mac.mm]: `notification.object` has type `id`, which doesn't necessarily point to an `NSString`. `base::SysNSStringToUTF8` will crash if you pass it an object that isn't an `NSString`. I'm somewhat surprised that this compiles, but in any case we should definitely check that the object is actually a string, e.g.

```objc
if ([notification.object isKindOfClass:[NSString class]]) {
  // ...
}
```

**Comment by MarshallOfSound** [comment]
The macOS codesign failure is due to the PR coming from a fork and is a known issue, if every other tests passes it's fine 👍

File: shell/browser/api/atom_api_system_preferences_mac.mm (hunk 2 above)
Comment (louismrose) [reply shell/browser/api/atom_api_system_preferences_mac.mm]: Thanks for the spotting this @nornagon, and for including sample code in your feedback. Made this a really easy fix for me!

**Comment by louismrose** [comment]
Thanks for the insightful feedback @nornagon. I've addressed the issue that you spotted.

CI is passing again now, except for the code-signing test which we can ignore. Thanks @MarshallOfSound for the tip about that one!

**Comment by louismrose** [comment]
Thanks very much for your help, everyone!

Dumb question -- how will I know when this change makes it into a release?

**Comment by nornagon** [comment]
You can watch release notes here: https://electronjs.org/releases

Current `master` will eventually become Electron 7, so the first stable version that will have this is 7.0.0.

**Comment by louismrose** [comment]
Perfect -- thanks for the tip!

Is there anything I could do to have this considered for a backport into Electron 6 (or even earlier)? This feature is a blocker for one of my employer's big summer projects, and we'd be happy to dedicate some time to a backport if appropriate.

**Comment by nornagon** [comment]
Unfortunately it's past the feature freeze time for Electron 6, so no new features will be accepted for backport. Once 6 is out though you'll be able to use the 7 beta series. See https://electronjs.org/docs/tutorial/electron-timelines.

**Comment by louismrose** [comment]
Ok, great -- we can likely make a beta series work for us. Thanks @nornagon!

**Pull request merged**